### PR TITLE
feat(shared-data, app, protocol-designer, api-client, api, robot-server): schemaV7 migration

### PR DESCRIPTION
**Pull request by jerader**
closes RAUT-515 and RAUT-272 and RQA-1128

# Overview

schemaV7 migration!

note: `LabwareLocation` type excludes `labwareId` object for now until we know more about the new `OnLabwareLocation` type Jeremy is making

coauthored with @TamarZanzouri 

# Test Plan

Smoke test the app, touchscreen, and Protocol-designer to make sure it is all working as expected.

PD sandbox: https://sandbox.designer.opentrons.com/shared_data-schemav7-migration/

# Changelog

- update the types, tests, components to work with schemaV7 types
- updated the `delay` command creator in step-generation to spit out `waitForResume` or `waitForDuration` since delay is deprecated
- created ` 7_0_0.ts` migration file since `pipettes`, `labware`, and `modules` key at the top level are deleted in schemaV7. Main changes are those keys are deleted and the `loadPipette`, `loadLabware`, and `loadModule` commands have updated params 

# Review requests

see test plan

# Risk assessment

touches a lot of code... so medium maybe

### CODE REVIEW

File: shared-data/protocol/types/schemaV7/command/setup.ts [hunk 1]
```diff
@@ -72,12 +72,16 @@ export type SetupCreateCommand =
   | LoadLiquidCreateCommand
   | MoveLabwareCreateCommand
 
-export type LabwareLocation = { slotName: string } | { moduleId: string }
+export type LabwareLocation =
+  | 'offDeck'
+  | { slotName: string }
+  | { moduleId: string }
```
Comment (shlokamin) [inline shared-data/protocol/types/schemaV7/command/setup.ts:79]: to incorporate @jbleon95's labware adapter split work this a labware location might also hold a labware id right?
```suggestion
  | { moduleId: string }
  | { labwareId: string }
```

Comment (jerader) [reply shared-data/protocol/types/schemaV7/command/setup.ts:79]: I spoke with brian about this last week and he said to worry about that at a follow up time - but happy to add it here if @jbleon95's work is merged in?

File: app/src/organisms/Devices/ProtocolRun/utils/getPickUpTipCommandsWithPipette.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
-import { PickUpTipRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV6/command/pipetting'
-import { RunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV6'
+import { PickUpTipRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV7/command/pipetting'
+import { RunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV7'
```
Comment (shlokamin) [inline app/src/organisms/Devices/ProtocolRun/utils/getPickUpTipCommandsWithPipette.ts]: not intro'd by this PR but these should be type imports

File: app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx [hunk 3]
```diff
@@ -150,7 +149,7 @@ export const LabwarePositionCheckComponent = (
         pipetteId: pip.id,
         labwareId: FIXED_TRASH_ID,
         wellName: 'A1',
-        wellLocation: { origin: 'default' as const },
+        wellLocation: { origin: 'top' as const },
```
Comment (shlokamin) [inline app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx:152]: i haven't been involved in schema interface change discussions but just wanna double check that this change here is intentional

File: app/src/organisms/OnDeviceDisplay/RunningProtocol/RunningProtocolCommandList.tsx [hunk 4]
```diff
@@ -173,7 +172,7 @@ export function RunningProtocolCommandList({
                   : COLORS.light1
               return (
                 <Flex
-                  key={command.id}
+                  key={command.key}
```
Comment (shlokamin) [inline app/src/organisms/OnDeviceDisplay/RunningProtocol/RunningProtocolCommandList.tsx]: this doesnt seem to matter but how come we're changing this from id to key?

File: app/src/organisms/RunPreview/index.tsx [hunk 5]
```diff
@@ -100,7 +100,7 @@ export const RunPreviewComponent = (
             : COLORS.darkGreyEnabled
           return (
             <Flex
-              key={command.id}
+              key={command.key}
```
Comment (shlokamin) [inline app/src/organisms/RunPreview/index.tsx]: same question here re: id vs key

File: app/src/resources/runs/utils.ts [hunk 6]
```diff
@@ -17,10 +17,13 @@ export const chainRunCommandsRecursive = (
...
     .then(response => {
-      if (!continuePastCommandFailure && response.data.status === 'failed') {
+      if (
+        !continuePastCommandFailure &&
+        response.data.result.status === 'failed'
+      ) {
         setIsLoading(false)
         return Promise.reject(
-          new Error(response.data.error?.detail ?? 'command failed')
+          new Error(response.data.result.error ?? 'command failed')
```
Comment (shlokamin) [inline app/src/resources/runs/utils.ts]: how come this change is necessary? does this fix a bug from a robot server interface change or something?

File: protocol-designer/src/steplist/fieldLevel/index.ts [hunk 7]
```diff
@@ -43,7 +43,7 @@ import {
   InvariantContext,
 } from '@opentrons/step-generation'
 import { StepFieldName } from '../../form-types'
-import { LabwareLocation } from '@opentrons/shared-data'
+import { LabwareLocation } from '@opentrons/shared-data/protocol/types/schemaV7/command/setup'
```
Comment (shlokamin) [inline protocol-designer/src/steplist/fieldLevel/index.ts]: i think this can be imported from the top level like it was before, and it can be typed prefixed!

File: shared-data/js/types.ts [hunk 8]
```diff
@@ -23,11 +23,11 @@ import {
   GRIPPER_V1_1,
   MAGNETIC_BLOCK_V1,
 } from './constants'
+import { EXTENSION } from '.'
```
Comment (shlokamin) [inline shared-data/js/types.ts]: not intro'd by this PR but we should be importing from constants rather than `.` to avoid the file importing itself

File: shared-data/protocol/types/schemaV7/command/index.ts [hunk 9]
```diff
@@ -40,17 +49,24 @@ export type CreateCommand =
...
-  // TODO (sb 10/26/22): Separate out calibration commands from protocol schema in RAUT-272
-  | CalibrationRunTimeCommand // for automatic pipette calibration
-  | AnnotationRunTimeCommand // annotating command execution
+  | ExtendedRunTimeCommand<PipettingRunTimeCommand> // involves the pipettes plunger motor
+  | ExtendedRunTimeCommand<GantryRunTimeCommand> // movement that only effects the x,y,z position of the gantry/pipette
+  | ExtendedRunTimeCommand<ModuleRunTimeCommand> // directed at a hardware module
+  | ExtendedRunTimeCommand<SetupRunTimeCommand> // only effecting robot's equipment setup (pipettes, labware, modules, liquid), no hardware side-effects
+  | ExtendedRunTimeCommand<TimingRunTimeCommand> // effecting the timing of command execution
+  | ExtendedRunTimeCommand<CalibrationRunTimeCommand> // for automatic pipette calibration
+  | ExtendedRunTimeCommand<AnnotationRunTimeCommand> // annotating command execution
```
Comment (shlokamin) [inline shared-data/protocol/types/schemaV7/command/index.ts]: don't runtime commands already extend `CommonCommandCreateInfo`?

File: shared-data/protocol/types/schemaV7/command/index.ts [hunk 10]
```diff
@@ -40,17 +49,24 @@ export type CreateCommand =
...
+
+interface RunCommandError {
+  id: string
+  errorType: string
+  createdAt: string
+  detail: string
+}
```
Comment (shlokamin) [inline shared-data/protocol/types/schemaV7/command/index.ts:70]: oh woa didnt know about this interface change. is this already in edge on the backend? if so we probably have bugs in the app right?

**Review by shlokamin — COMMENTED** [review]
code lgtm, need some time to test on a robot tho

File: app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx (hunk 3 above)
Comment (jerader) [reply app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx:152]: `default` isn't a well location origin option in schema v7, i saw a comment somewhere saying that the new `default` is `top`. (can't remember where i saw that comment anymore though)

File: shared-data/protocol/types/schemaV7/command/index.ts (hunk 10 above)
Comment (jerader) [reply shared-data/protocol/types/schemaV7/command/index.ts:70]: just copied over what was in schemav6 because i had some type error somewhere that needed it. i'm assuming it is in edge on the backend.

File: app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx (hunk 3 above)
Comment (jerader) [reply app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx:152]: oh just found it: `// default value is 'top'` is a comment in the `WellLocationParams` type in `shared-data/protocol/types/schemaV7/command/pipetting.ts`

File: app/src/organisms/OnDeviceDisplay/RunningProtocol/RunningProtocolCommandList.tsx (hunk 4 above)
Comment (jerader) [reply app/src/organisms/OnDeviceDisplay/RunningProtocol/RunningProtocolCommandList.tsx]: good catch idk why i changed it actually

File: app/src/resources/runs/utils.ts (hunk 6 above)
Comment (jerader) [reply app/src/resources/runs/utils.ts]: oops idk why i changed this, good catch!

File: shared-data/protocol/types/schemaV7/command/index.ts (hunk 9 above)
Comment (jerader) [reply shared-data/protocol/types/schemaV7/command/index.ts]: i don't think so? I get a bunch of check errors saying properties that are part of `CommonCommandRunTimeInfo` don't exist when i don't extend the types like this.

**Comment by shlokamin** [comment]
@jerader i believe you're already aware (i just remembered/realized), but when importing this PD protocol i made on your branch into the app i get a protocol analysis error:  
```
Error: schema v7 migration test.json could not be read as a JSON protocol.

```

zipping the file so i can upload it here to github: 
[schema v7 migration test.json.zip](https://github.com/Opentrons/opentrons/files/11983537/schema.v7.migration.test.json.zip)

this makes sense, because the app is using the opentrons analyze cli which doesn't yet know about schema v7. this also means we can't really test these changes at all, and that if we merge this into edge and cut another internal robot stack release, this will probably break LPC and other flows.

sooooo i don't think we can merge this as is. we either need the backend changes, or we only merge PD/step generation changes with a follow up PR coming soon to fix the backend stuff. 

when did RSS say that can make the backend changes?

**Comment by shlokamin** [comment]
seeing this LPC error on a robot when running a python protocol:

Screenshot 2023-07-07 at 12 09 14 PM


looks like an issue with chain run commands. i think it's this line in `app/src/resources/runs/utils.ts`

`response.data.result.status === 'failed'`

File: shared-data/protocol/types/schemaV7/command/index.ts (hunk 9 above)
Comment (shlokamin) [reply shared-data/protocol/types/schemaV7/command/index.ts]: hmm but if you look at the runtime commands, it looks like the extend `CommonCommandCreateInfo`:

```js
export interface MoveLabwareRunTimeCommand
  extends CommonCommandRunTimeInfo,
    MoveLabwareCreateCommand {
  result?: MoveLabwareResult
}
```

File: shared-data/protocol/types/schemaV7/command/index.ts (hunk 10 above)
Comment (shlokamin) [reply shared-data/protocol/types/schemaV7/command/index.ts:70]: so the backend needs to change to reflect this interface change? i dont think what the server currently returns reflects this. i'd imagine that's what RSS is gonna do tho

**Comment by jerader** [comment]
> looks like an issue with chain run commands. i think it's this line in `app/src/resources/runs/utils.ts`
> 
> `response.data.result.status === 'failed'`

@shlokamin  Oops ya i see the issue, thank you! should be `response.data.status === 'failed'`

File: shared-data/protocol/types/schemaV7/command/index.ts (hunk 9 above)
Comment (jerader) [reply shared-data/protocol/types/schemaV7/command/index.ts]: omg i see what happened, one of the commands didn't extend it. fixing it now!

File: app/src/organisms/Devices/ProtocolRun/SetupLabwarePositionCheck/__tests__/CurrentOffsetsTable.test.tsx [hunk 11]
```diff
@@ -83,24 +83,48 @@ describe('CurrentOffsetsTable', () => {
...
+        definitionId: 'opentrons/opentrons_96_tiprack_300ul/1',
+      },
+      '9fbc1db0-0042-11ec-8258-f7ffdf5ad45a:opentrons/nest_12_reservoir_15ml/1': {
+        displayName: 'NEST 12 Well Reservoir 15 mL',
+        definitionId: 'opentrons/nest_12_reservoir_15ml/1',
+      },
+      'e24818a0-0042-11ec-8258-f7ffdf5ad45a': {
+        displayName: 'Opentrons 96 Tip Rack 300 µL (1)',
+        definitionId: 'opentrons/opentrons_96_tiprack_300ul/1',
+      },
```
Comment (koji) [inline app/src/organisms/Devices/ProtocolRun/SetupLabwarePositionCheck/__tests__/CurrentOffsetsTable.test.tsx:122]: question
This one is needed? Looks pretty similar to  L111-113.
or testing `(1)` part?

File: app/src/organisms/Devices/ProtocolRun/utils/getProtocolModulesInfo.ts [hunk 12]
```diff
@@ -3,7 +3,7 @@ import {
   getModuleDef2,
   getLoadedLabwareDefinitionsByUri,
 } from '@opentrons/shared-data'
-import { LoadLabwareRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV6/command/setup'
+import { LoadLabwareRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV7/command/setup'
```
Comment (koji) [inline app/src/organisms/Devices/ProtocolRun/utils/getProtocolModulesInfo.ts]: maybe `import type` since path is  types/schemaV7 ?

File: app/src/organisms/Devices/ProtocolRun/utils/getTipracksVisited.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import { PickUpTipRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV6/command/pipetting'
+import { PickUpTipRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV7/command/pipetting'
```
Comment (koji) [inline app/src/organisms/Devices/ProtocolRun/utils/getTipracksVisited.ts]: same maybe `import type`?

File: app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx [hunk 14]
```diff
@@ -27,12 +31,7 @@ import {
 } from '@opentrons/shared-data'
 import type { Axis, Sign, StepSize } from '../../molecules/JogControls/types'
 import type { RegisterPositionAction, WorkingOffset } from './types'
-import { LabwareOffsetCreateData } from '@opentrons/api-client'
-import { getLabwarePositionCheckSteps } from './getLabwarePositionCheckSteps'
-import { DropTipCreateCommand } from '@opentrons/shared-data/protocol/types/schemaV6/command/pipetting'
-import { useChainMaintenanceCommands } from '../../resources/runs/hooks'
-import { FatalErrorModal } from './FatalErrorModal'
-import { RobotMotionLoader } from './RobotMotionLoader'
+import type { DropTipCreateCommand } from '@opentrons/shared-data/protocol/types/schemaV7/command/pipetting'
```
Comment (koji) [inline app/src/organisms/LabwarePositionCheck/LabwarePositionCheckComponent.tsx]: Can you put this line after `import type { LabwareOffset } from '@opentrons/api-client'` and move up L27-L31?

File: app/src/organisms/ModuleCard/TemperatureModuleSlideout.tsx [hunk 15]
```diff
@@ -18,7 +18,7 @@ import { Slideout } from '../../atoms/Slideout'
 import { SubmitPrimaryButton } from '../../atoms/buttons'
 import { InputField } from '../../atoms/InputField'
 import { StyledText } from '../../atoms/text'
-import { TemperatureModuleSetTargetTemperatureCreateCommand } from '@opentrons/shared-data/protocol/types/schemaV6/command/module'
+import { TemperatureModuleSetTargetTemperatureCreateCommand } from '@opentrons/shared-data/protocol/types/schemaV7/command/module'
```
Comment (koji) [inline app/src/organisms/ModuleCard/TemperatureModuleSlideout.tsx]: `import type`?

File: app/src/organisms/ProtocolDetails/__tests__/RobotConfigurationDetails.test.tsx [hunk 16]
```diff
@@ -5,7 +5,7 @@ import { OT2_STANDARD_MODEL, OT3_STANDARD_MODEL } from '@opentrons/shared-data'
 import { i18n } from '../../../i18n'
 import { useFeatureFlag } from '../../../redux/config'
 import { RobotConfigurationDetails } from '../RobotConfigurationDetails'
-import { LoadModuleRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV6/command/setup'
+import { LoadModuleRunTimeCommand } from '@opentrons/shared-data/protocol/types/schemaV7/command/setup'
```
Comment (koji) [inline app/src/organisms/ProtocolDetails/__tests__/RobotConfigurationDetails.test.tsx]: `import type` ?

File: protocol-designer/src/file-data/__fixtures__/createFile/v7Fixture.ts [hunk 17]
```diff
@@ -0,0 +1,315 @@
+// Named arguments to createFile selector. This data would be the result of several selectors.
+import { RobotState, Timeline } from '@opentrons/step-generation'
+import { SavedStepFormState, ModuleEntities } from '../../../step-forms'
+import { StepIdType } from '../../../form-types'
```
Comment (koji) [inline protocol-designer/src/file-data/__fixtures__/createFile/v7Fixture.ts]: maybe `import type`?

File: app/src/organisms/Devices/ProtocolRun/SetupLabwarePositionCheck/__tests__/CurrentOffsetsTable.test.tsx (hunk 11 above)
Comment (jerader) [reply app/src/organisms/Devices/ProtocolRun/SetupLabwarePositionCheck/__tests__/CurrentOffsetsTable.test.tsx:122]: oh its a 2nd tiprack in the protocol that is the same type as the 1st tiprack. i kept the labware the same here as the fixture it used to use

File: shared-data/protocol/types/schemaV7/command/index.ts [hunk 18]
```diff
@@ -15,6 +15,14 @@ import type {
   CalibrationCreateCommand,
 } from './calibration'
 
+export * from './pipetting'
+export * from './gantry'
+export * from './module'
+export * from './setup'
+export * from './timing'
+export * from './annotation'
+export * from './calibration'
```
Comment (koji) [inline shared-data/protocol/types/schemaV7/command/index.ts]:
```suggestion
export * from './annotation'
export * from './calibration'
export * from './gantry'
export * from './module'
export * from './pipetting'
export * from './setup'
export * from './timing'
```

File: step-generation/src/fixtures/commandFixtures.ts [hunk 19]
```diff
@@ -3,7 +3,7 @@ import type {
   AspDispAirgapParams,
   BlowoutParams,
   TouchTipParams,
-} from '@opentrons/shared-data/protocol/types/schemaV6/command/pipetting'
+} from '@opentrons/shared-data/protocol/types/schemaV7/command/pipetting'
 import { FIXED_TRASH_ID } from '../constants'
```
Comment (koji) [inline step-generation/src/fixtures/commandFixtures.ts]: Can you move up this import statement?

File: shared-data/protocol/schemas/7.json [hunk 20]
```diff
@@ -18,7 +18,12 @@
         "p300_single_gen2",
         "p300_multi_gen2",
         "p1000_single",
-        "p1000_single_gen2"
+        "p1000_single_gen2",
+        "p50_single_gen3",
+        "p50_multi_gen3",
+        "p1000_single_gen3",
+        "p1000_multi_gen3",
+        "p1000_96"
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/7.json]: Just making sure: These changes are intentional, right? And it's right for their names to say "gen3" instead of "flex"?

**Review by SyntaxColoring — APPROVED** [review]
Python and schema changes look good to me. Thanks!

File: shared-data/protocol/schemas/7.json (hunk 20 above)
Comment (smb2268) [reply shared-data/protocol/schemas/7.json]: I think these should be `flex` not `gen3`

Comment (jerader) [reply shared-data/protocol/schemas/7.json]: oh good catch! should probably be updated to say `flex`

File: api/src/opentrons/protocol_runner/json_translator.py [hunk 21]
```diff
@@ -87,8 +143,27 @@ def _translate_pipette_command(
     return translated_obj
 
 
+def _translate_v7_pipette_command(
+    command: protocol_schema_v7.Command,
+) -> pe_commands.CommandCreate:
+    pipette_id = command.params.pipetteId
+    # v7 data model supports all commands and therefor most props are optional.
+    # load pipette command must contain pipette_id.
+    assert pipette_id is not None
```
Comment (sfoster1) [inline api/src/opentrons/protocol_runner/json_translator.py:152]: nit: can these be enumerated errors?

**Review by b-cooper — APPROVED** [review]
🌟
